### ISSUE TITLE
I update to 0.5.4, error on MODULE_NOT_FOUND

### ISSUE THREAD

**Issue by ncaq** [getsentry/sentry-electron#81]
I update `@sentry/electron` 0.4.1 to 0.5.4.

And I rewrite function create to init.

My `src/sentry.js`.

~~~
import { init, captureMessage } from '@sentry/electron';

init({ dsn: 'foobar' });
captureMessage('Hello, world!');
~~~

So sentry is failed.
(My application name replaced to foo)

~~~
{ Error: Cannot find module "/home/ncaq/Desktop/foo/package.json".
    at webpackEmptyContext (/home/ncaq/Desktop/foo/app/webpack:/node_modules/@sentry/electron/dist/main sync:2:1)
    at getPackageJson (/home/ncaq/Desktop/foo/app/webpack:/node_modules/@sentry/electron/dist/main/context.js:80:1)
    at /home/ncaq/Desktop/foo/app/webpack:/node_modules/@sentry/electron/dist/main/context.js:240:25
    at Generator.next (<anonymous>)
    at fulfilled (/home/ncaq/Desktop/foo/app/webpack:/node_modules/@sentry/electron/dist/main/context.js:4:42)
    at <anonymous> code: 'MODULE_NOT_FOUND' }
~~~

However, `/home/ncaq/Desktop/foo/package.json` exist.
Since this is a full path, it is hard to think that searching for webpack or babel is the cause.

Is it bug?
Or my setting is invalid?

---

My env.

* babel: 7.0.0-beta.46
* `@sentry/electron`: 0.5.4
* react: 16.2.0
* webpack: 4.6.0
* electron: 1.8.4
* OS: Gentoo Linux
* node: v9.10.1

**Comment by jan-auer** [comment]
The reason for this error is that webpack replaces `require`, which is required by the SDK to load your app's package.json. Can you please try out the fix from #82 and see whether it works now?

```bash
npm install 'https://github.com/getsentry/sentry-electron#fix/webpack-require'
```

**Comment by maplemike** [comment]
I followed the instructions and get the following error now instead of module not found:

>TypeError: e.require is not a function
    at a (/xxx/app/webpack:/sentry-electron/dist/main/context.js:83:25)
    at /xxx/app/webpack:/sentry-electron/dist/main/context.js:2:45
    at Generator.next (<anonymous>)
    at i (/xxx/app/webpack:/sentry-electron/dist/main/context.js:6:33)
    at <anonymous>

**Comment by jan-auer** [comment]
Could you please help me out by sharing the relevant bits of your webpack configuration, so I can reproduce this in a test case? Thanks!

**Comment by dimitarnestorov** [comment]
I also use webpack and what I've done is added the package in my config to `externals` like so

```javascript
{
	externals: {
		'@sentry/electron': 'require("@sentry/electron")'
	}
}
```

[webpack documentation on `externals`](https://webpack.js.org/configuration/externals/)

-----------------
In case you don't like it/doesn't work I have an alternative:

```javascript
const { init, captureMessage } = window.require('@sentry/electron');

init({ dsn: 'foobar' });
captureMessage('Hello, world!');
```

**Comment by jan-auer** [comment]
Wait, are you getting this error in a renderer?

**Comment by dimitarnestorov** [comment]
Well why would you use webpack in the main process?

**Comment by maplemike** [comment]
Downgrading to 0.5.0 is the oldest version I could get to work properly. My webpack config files are here: https://gist.github.com/maplemike/4a146236a8bf9eac2929670ec1d8f35c

Everything works fine in dev, but in production the module not found error seems to creep up. I'm not too familiar with webpack and Elecron so not sure where to start here

**Comment by jan-auer** [comment]
I have seen people do that, that's why I was so irritated. This code section should not even be executed in a renderer due to [this dispatch](https://github.com/getsentry/sentry-electron/blob/master/src/dispatch.ts#L21-L24). I suppose that some optimization removes this code from your bundles?

**Comment by jan-auer** [comment]
Can you try to use [WebpackDefinePlugin](https://webpack.js.org/plugins/define-plugin/) to set `process.type` to `"renderer"`?

**Comment by maplemike** [comment]
adding this to the renderer webpack config

>new webpack.DefinePlugin({
      'process.type': JSON.stringify('renderer')
    }),

had no effect

**Comment by jan-auer** [comment]
Next attempt in #84, would someone please be so kind to try this out? 

I also added webpack to the example app this time, as we'll need it for the sandbox mode soon, anyway.

**Comment by maplemike** [comment]
I still get an error...

>App threw an error during load
TypeError: e.require is not a function
    at new t.ElectronFrontend (/xxx/app/webpack:/sentry-electron/dist/dispatch.js:37:9)

**Comment by jan-auer** [comment]
@maplemike can you please confirm that the error is being thrown in a renderer and your webpack configuration does not override `process.type` anymore? I'm having a hard time reproducing this error. 

If you're in the main process, you'll have to make sure that `@sentry/electron` is specified in the `externals` option of your webpack config.

**Comment by maplemike** [comment]
@jan-auer apologies I'm fairly new to Javascript + Electron/Webpack so I've yet to be able to dive deeper and research this

* all of my dependencies are already in externals in webpack config
* process.type is 'browser' in the main process when I did a console.log (overriding it to 'renderer' causes an exception)
* my renderer throws an error which is uncaught
* the uncaught error seems to bubble up to the sentry instance in the main process, even though I have a sentry instance defined in the renderer process as well - I thought before with raven-js this didn't happen but perhaps I'm mistaken

Does this help at all? If not I will try to produce a minimal example that causes this to happen

**Comment by jan-auer** [comment]
@maplemike The Electron SDK in the renderer is just a proxy to the main process, so all events, context and breadcrumbs are sent there. This is why you will see your renderer errors bubble up to the main process Sentry instance. It's totally expected.

From your description I guess that the `TypeError` you posted above is being thrown by the Sentry instance in the main process? This makes me wonder whether you also use webpack for the main process or only for the renderer?

Here's my take on it right now, based on the example app:
 1. In the renderer process you can use webpack without any limitations, except you have to configure `electron` as external. The SDK should work and send all events to the main process.
 2. In the main process, I have never tested webpack but it should work. However, you need to make sure that `@sentry/electron` is an external. If webpack in the main process does not work (and generate the above error instead), I will try to come up with a better solution.

**Comment by ncaq** [comment]
I try `fix/dispatch` branch.

I see new error.

~~~
yarn run v1.6.0
$ cross-env NODE_ENV=development electron .
App threw an error during load
TypeError: module.require is not a function
    at new ElectronFrontend (/home/ncaq/Desktop/foo/app/webpack:/sentry-electron/dist/dispatch.js:38:22)
    at Object.initAndBind (/home/ncaq/Desktop/foo/app/webpack:/node_modules/@sentry/core/dist/sdk.js:24:1)
    at init (/home/ncaq/Desktop/foo/app/webpack:/sentry-electron/dist/sdk.js:51:12)
    at Object../src/sentry.js (/home/ncaq/Desktop/foo/app/webpack:/src/sentry.js:5:1)
    at __webpack_require__ (/home/ncaq/Desktop/foo/app/webpack:/webpack/bootstrap:19:1)
    at Object.<anonymous> (/home/ncaq/Desktop/foo/app/webpack:/src/main.js:17:1)
    at Object../src/main.js (/home/ncaq/Desktop/foo/app/main.js:61822:30)
    at __webpack_require__ (/home/ncaq/Desktop/foo/app/webpack:/webpack/bootstrap:19:1)
    at ../sentry-electron/dist/common.js.Object.defineProperty.value (/home/ncaq/Desktop/foo/app/webpack:/webpack/bootstrap:68:1)
    at Object.<anonymous> (/home/ncaq/Desktop/foo/app/main.js:72:10)
A JavaScript error occurred in the main process
Uncaught Exception:
TypeError: module.require is not a function
    at new ElectronFrontend (/home/ncaq/Desktop/foo/app/webpack:/sentry-electron/dist/dispatch.js:38:22)
    at Object.initAndBind (/home/ncaq/Desktop/foo/app/webpack:/node_modules/@sentry/core/dist/sdk.js:24:1)
    at init (/home/ncaq/Desktop/foo/app/webpack:/sentry-electron/dist/sdk.js:51:12)
    at Object../src/sentry.js (/home/ncaq/Desktop/foo/app/webpack:/src/sentry.js:5:1)
    at __webpack_require__ (/home/ncaq/Desktop/foo/app/webpack:/webpack/bootstrap:19:1)
    at Object.<anonymous> (/home/ncaq/Desktop/foo/app/webpack:/src/main.js:17:1)
    at Object../src/main.js (/home/ncaq/Desktop/foo/app/main.js:61822:30)
    at __webpack_require__ (/home/ncaq/Desktop/foo/app/webpack:/webpack/bootstrap:19:1)
    at ../sentry-electron/dist/common.js.Object.defineProperty.value (/home/ncaq/Desktop/foo/app/webpack:/webpack/bootstrap:68:1)
    at Object.<anonymous> (/home/ncaq/Desktop/foo/app/main.js:72:10)
~~~

I check `process.type` is, in electron-main `browser`, in electron-render `renderer`.

I did not set `@sentry/electron` to external.
I try set it, But error.

**Comment by aguynamedben** [comment]
@maplemike My project is also based on electron-react-boilerplate and uses the two-package.json structure. I'm having all kinds of issues getting this working.

When you got it working, were you using the latests release on npm or were you using the fix/dispatch branch?

Currently getting
```
"Uncaught Error: Cannot find module 'require("@sentry/electron")'".
```

I put this in my webpack.config.base.js:

```js
import { dependencies as externals } from './app/package.json';

const externalsObj = Object.keys(externals || {}).reduce((obj, key) => {
  obj[key] = key;
  return obj;
}, {});
externalsObj['@sentry/electron'] = 'require("@sentry/electron")';
console.log(`externalsObj is`, externalsObj);

export default {
  // Required for node-notifier to work
  // https://www.npmjs.com/package/node-notifier#using-webpack
  node: {
    __filename: true,
    __dirname: true,
  },

  externals: externalsObj,
// and so on ...
```

Log output showing what externals is set to:
```
externalsObj is { pouchdb: 'pouchdb',
  '@sentry/electron': 'require("@sentry/electron")' }
```

I also tried adding `@sentry/electron` to app/package.json because I noticed webpack.config.base.js adds dependencies from app/package.json to externals, and everything seemed to work, but when my source maps were uploaded there were issues. (Happy to share the Sentry issues with employees)

The electron-react-boilderplate is very popular, it would be a great real-world proof-of-concept to get this working in that boilerplate.

**Comment by maplemike** [comment]
@aguynamedben This was a nightmare to get going...I'm a Python developer and have been using Sentry for years

webpack.config.base.js
```javascript
import path from 'path';
import webpack from 'webpack';
import { dependencies as externals } from './app/package.json';

export default {
  externals: Object.keys(externals || {}),
// ... etc
```
### What I did...
- I removed and uninstalled @sentry/electron from my main package.json/node_modules (it would still fail before I did this)
- Went into the app directory and installed @sentry/electron
- Using the current latest release not the fix/dispatch branch

This resolved the module not found error, source maps were another issue...

### Sourcemaps
Updating the latest version of the boilerplate should fix this, it was caused by the boilerplate producing bad sourcemaps

If that is too much:
- Uninstall babili-webpack-plugin and remove it from your webpack configurations
- Install uglifyjs-webpack-plugin
- Under plugins (for both webpack.config.main.prod.js and webpack.config.renderer.prod.js ) you'll want to add:

```javascript
new UglifyJSPlugin({
      parallel: true,
      sourceMap: true
    }),
```

Hope that helps...finally have it all working!

**Comment by aguynamedben** [comment]
@maplemike Thanks so much. I landed on something similar. What worked for me with a project based on electron-react-boilerplate and Electron 1.7.14.

#### Use app/package.json in the two-package.json scheme
Add @sentry/electron to app/package.json, not package.json. I think this makes sure that the package is in Webpack `externals` as directed by @jan-auer.

```
cd app
yarn add @sentry/electron
cd ..
```

#### In main process (app/main.dev.js)

```js
import {
  app,
  BrowserWindow,
  ipcMain,
  globalShortcut,
  Notification,
  crashReporter,
  Menu,
} from 'electron';
import * as Sentry from '@sentry/electron';

// Setup Electron crashReporter
const crashReporterOptions = {
  companyName: 'My Company',
  productName: 'My Product',
  ignoreSystemCrashHandler: true,
  submitURL: process.env.SENTRY_CRASH_REPORT_URL,
};
crashReporter.start(crashReporterOptions);

// Sentry setup (Electron SDK)
Sentry.init({
  dsn: process.env.SENTRY_DSN,
  release: `${process.env.SENTRY_PROJECT}-${app.getVersion()}`,
});

// When my main process receives credentials from the renderer process (user logged in)
ipcMain.on('receive-credentials', (event, credentials) => {
  // Update crashReporter (WARN: macOS/Windows/Linux APIs differ)
  if (process.platform === 'darwin') {
    crashReporter.setExtraParameter('user_id', credentials.id);
    crashReporter.setExtraParameter('user_email', credentials.email);
  } else {
    crashReporter.start({
      ...crashReporterOptions,
      user_id: credentials.id,
      user_email: credentials.email
    });
  }

  Sentry.setUserContext({
    id: credentials.id,
    email: credentials.email,
  });
});
```

#### In renderer process (app/index.js)

```js
import * as Sentry from '@sentry/electron';

Sentry.init({
  dsn: process.env.SENTRY_DSN,
  release: `${process.env.SENTRY_PROJECT}-${remote.app.getVersion()}`,
});
```

#### Script for uploading sourcemaps with sentry-cli

I gave up on using the Webpack plugin since that is confused by electron-react-boilerplate having 2 different Webpack setups (1 for main process, 1 for renderer process). Instead, I just use sentry-cli.

Add sentry-cli to devDependencies in top-level package.json
```
yarn add --dev @sentry/cli
```

upload_sourcemaps.sh
```sh
#!/bin/bash

echo "Uploading source maps to Sentry"

ENV=$1

if [ -z "$ENV" ] ; then
  echo "No argument supplied, please pass the environment"
  exit 1
elif [ "$ENV" != "development" ] && [ "$ENV" != "production" ] ; then
  echo "Environment must be 'development' or 'production', not '$ENV'"
  exit 1
fi

echo "Environment: $ENV"

ENV_FILE_PATH="./env.$ENV"
echo "env file is $ENV_FILE_PATH"
source "$ENV_FILE_PATH"

echo "SENTRY_AUTH_TOKEN: $SENTRY_AUTH_TOKEN"
echo "SENTRY_PROJECT: $SENTRY_PROJECT"
echo "SENTRY_ORG: $SENTRY_ORG"

export SENTRY_AUTH_TOKEN=$SENTRY_AUTH_TOKEN
export SENTRY_PROJECT=$SENTRY_PROJECT
export SENTRY_ORG=$SENTRY_ORG

# https://gist.github.com/DarrenN/8c6a5b969481725a4413#gistcomment-1678696
PACKAGE_VERSION=$(cat app/package.json \
  | grep version \
  | head -1 \
  | awk -F: '{ print $2 }' \
  | sed 's/[",]//g' \
  | tr -d '[[:space:]]')

echo "Package version: $PACKAGE_VERSION"

echo "Uploading..."
./node_modules/.bin/sentry-cli releases files $SENTRY_PROJECT-$PACKAGE_VERSION upload-sourcemaps app/dist --url-prefix '~/dist/'
```

#### In the top-level package.json

Change `yarn release` command:
```
    "release": "yarn && npm run build && ./upload_sourcemaps.sh production && build",
```

I also added this command so I could easily test by tweaking my env.production (I use dotenv to manage environmental variables)
```
    "upload-sourcemaps-production": "./upload_sourcemaps.sh production",
```

With all this setup, running `yarn release` to ship a new version of the app uploads the sourcemaps to Sentry as a release called `myapp-production-1.1.0` before uploading the fully built macOS package to Amazon S3. The release is set similarly in the code. Errors are coming in with sourcemaps working. 👍

**Comment by timfish** [comment]
@aguynamedben You don't need the start the native crash reporter manually. 
`@sentry/electron` already configures this for you in [the main](https://github.com/getsentry/sentry-electron/blob/master/src/main/backend.ts#L230) and [renderer](https://github.com/getsentry/sentry-electron/blob/master/src/renderer/backend.ts#L113) processes.

**Comment by fangzhzh** [comment]
one small thing to add to @aguynamedben's answer about the sourcemaps upload: 
following the [document](https://docs.sentry.io/clients/electron/sourcemaps/)

add following snippet to your webpack.**.prod.js
```
const SentryWebpackPlugin = require('@sentry/webpack-plugin');

module.exports = {
  // other configuration
  plugins: [
    new SentryWebpackPlugin({
      include: 'app',
      ignoreFile: '.sentrycliignore',
      ignore: ['node_modules', 'webpack.config.js'],
      configFile: 'sentry.properties'
    })
  ]
};
```
it will upload for you when you make a release build

**Comment by jan-auer** [comment]
Seems like electron-react-boilerplate is causing some confusion with their multiple package.jsons and configuration. I have to admit that I haven't tried that boilerplate yet, but will give it a go and see whether we can make the integration process easier. Thanks for all the feedback so far. 

Having said that, we can probably not account for the seemingly complex way the boilerplate is structured. My goal is to get this SDK to work with webpack, and not this specific boilerplate.

Some general things:
 - You should include `@sentry/electron` in both the main process and renderer javascripts in the exact same way. In the main process webpack configuration, it should be declared external. In renderer  webpack configuration it should **not** be declared external (although it wouldn't matter as long as you have node integration enabled).
 - SentryWebpackPlugin should work in any case. If something does not work as expected, please file an issue [here](https://github.com/getsentry/sentry-webpack-plugin/issues)
 - There is no need to start the Electron crash reporter manually, as @timfish has pointed out. The SDK does this when you initialize it (and also takes care of platform specifics for you)

**Comment by jan-auer** [comment]
Just tried out `electron-react-boilerplate`, it seems pretty straight-forward if you pay attention to their readme and this thread (`externals`!). The way this boilerplate is structured, you have to add it inside the `app/` directory, which is what @maplemike already pointed out.

**Comment by jan-auer** [comment]
We've just released `v0.5.5` and updated our documentation accordingly. I hope this solves your issues - please let me know if you're still experiencing troubles.

**Comment by jan-auer** [comment]
I'm closing this since this has essentially been fixed with the release of v0.5.5 and updated documentation. Many thanks to everyone participating in this discussion and providing so much input. 

Feel free to reply if you are still running into issues.

**Comment by aguynamedben** [comment]
Thanks @jan-auer.

As you may have seen, the seemingly complex [two package.json structure](https://github.com/chentsulin/electron-react-boilerplate/wiki/Module-Structure----Two-package.json-Structure) of the boilerplate is so that people can include native dependencies in their Electron project... i.e. a C++ module for a LevelDB/sqlite/etc. binding. It's definitely a non-trivial but core need for many apps. For example, my app is using [PouchDB](https://github.com/pouchdb/pouchdb)+LevelDB to store data.

Thanks for taking the time to check out the boilerplate. It's one of the main Electron boilerplate projects, so I think it should be treated as a good measuring stick for an Electron SDK.

The docs for this SDK are great, thanks for updating them. 👏
